visualize-admin/visualization-tool
# feat: Introduce component ids

Closes #1696

This PR introduces a new, **component id** concept to Visualize, which replaces the current _component iri_ convention. 

**Component id** is made by combining `unversioned cube iri` with `unversioned component iri`, which is a unique combination even in case we merged two cubes with the same `component iris`. This means that we can handle these cases correctly, while leaving our business logic mostly intact (and in some cases, even simpler, as we don't need to filter by `cube iri` and `component iri` anymore, but instead simply use **component id**).

This combination of `iris` is done on the server side and sent to the client, which doesn't need to know about this fact, but simply keep the current logic (chart config migrations are the only exception, as we need to introduce `unversioned cube iris` to relevant parts of the config dynamically to make sure old charts work).

By proxy, it was necessary to make the migrations async, so we can fetch `unversioned cube iri` when needed – as [immer doesn't support / recommend asynchronous operations](https://immerjs.github.io/immer/async/), I removed it.

ℹ️ [Dimension id docs](https://www.notion.so/interactivethings/Dimensions-87f4b91b502045ba8f1486439fdb1691?pvs=4#14430da54302800087f1ef1364b23d55)

🔎 [Visual regression testing docs](https://www.notion.so/interactivethings/Visual-regression-testing-14430da5430280cb83e0d0ccbe657a88)

### How to test
🚨 **Preferably, the implementation should be tested holistically, as essentially every part of Visualize was affected in one or another way.**

1. Go to [this link](https://visualization-tool-git-feat-finalize-merging-of-cubes-2-ixt1.vercel.app/en/create/new?cube=https://agriculture.ld.admin.ch/foag/cube/MilkDairyProducts/Consumption_Price_Month&dataSource=Prod).
2. Click on `Add dataset`.
3. Add `Market figures milk and dairy products - Monthly producer prices` dataset.
4. Open `Vertical axis`.
5. ✅ See that it's possible to toggle between two `Price` dimensions that have the same `iri`s and names, but come from different cubes.
6. Change chart type to `Table`.
7. ✅ See that there are duplicated dimensions visible in the `Table` (that have the same `iri`s and names, but come from different cubes).

## How to reproduce
Repeat the above steps on PROD environment and see that ✅ points are broken ❌.

---
- [x] Add docs to feature pages in Notion (both for `componentId` and visual regression testing)
- [x] Check E2E tests locally
- [x] Check if all old charts work (checked local `/_charts` page and cross-compared 500 PROD charts on `main` vs this branch)
